fcrepo-clients/src/main/java/edu/unc/lib/dl/fcrepo4/RepositoryObject.java

```diff
@@ -0,0 +1,196 @@
+/**
+ * Copyright 2016 The University of North Carolina at Chapel Hill
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.unc.lib.dl.fcrepo4;
+
+import java.net.URI;
+import java.util.Date;
+import java.util.List;
+
+import com.hp.hpl.jena.rdf.model.Model;
+
+import edu.unc.lib.dl.fedora.FedoraException;
+import edu.unc.lib.dl.fedora.PID;
+
+/**
+ * A generic repository object, with properties common to objects in the repository.
+ * 
+ * @author bbpennel
+ *
+ */
+public abstract class RepositoryObject {
+
+	// Repository which produced and manages this object
+	protected Repository repository;
+	// Loader for lazy loading data about this object when requested
+	protected RepositoryObjectDataLoader dataLoader;
+
+	// The identifier and path information for this object
+	protected PID pid;
+
+	protected Model model;
+
+	protected Date lastModified;
+	protected String etag;
+
+	protected List<String> types;
+
+	protected RepositoryObject(PID pid, Repository repository, RepositoryObjectDataLoader dataLoader) {
+		this.repository = repository;
+		this.pid = pid;
+		this.dataLoader = dataLoader;
+	}
+
+	/**
+	 * Get the URI of this object in the repository
+	 * 
+	 * @return
+	 */
+	public URI getUri() {
+		return pid.getRepositoryUri();
+	}
+
+	/**
+	 * Get a model containing properties and relations held by this object. 
+	 * 
+	 * @return
+	 * @throws FedoraException
+	 */
+	public Model getModel() throws FedoraException {
+		if (model == null) {
+			dataLoader.loadModel(this);
+		}
+
+		return model;
+	}
+
+	/**
+	 * Set the model
+	 * 
+	 * @param model
+	 * @return
+	 */
+	public RepositoryObject setModel(Model model) {
```

Comment:
Geneally it would be to allow for method chaining (ie, obj.setModel(...).setBlah(...).doSomething()), but I haven't been consistent about it here and it doesn't seem like that useful of a pattern for RepositoryObjects and their descendents, so I am going to switch to void for these methods